agents/cache/cache.go

```diff
@@ -0,0 +1,194 @@
+type statementsMap map[interface{}]interface{}
+type statementsAddedMap map[interface{}]time.Time
+
+// Cache provides cached access to performance statistics tables.
+// It retains data longer than those tables.
+// Intended to store various subtypes of map.
+type Cache struct {
+	typ       reflect.Type
+	items     statementsMap
+	added     statementsAddedMap
+	retain    time.Duration
+	sizeLimit uint
+	l         *logrus.Entry
+	rw        sync.RWMutex
+	updatedN  uint
+	addedN    uint
+	removedN  uint
+}
+
+// New creates new Cache.
+// Argument typ is an instance of type to be stored in Cache, must be a map with chosen key and value types.
+func New(typ interface{}, retain time.Duration, sizeLimit uint, l *logrus.Entry) (*Cache, error) {
+	if reflect.TypeOf(typ).Kind() != reflect.Map {
+		return nil, fmt.Errorf("%w: typ must be of map kind", ErrWrongType)
+	}
+	return &Cache{
+		typ:       reflect.TypeOf(typ),
+		retain:    retain,
+		sizeLimit: sizeLimit,
+		l:         l,
+		items:     make(statementsMap),
+		added:     make(statementsAddedMap),
+	}, nil
+}
+
+// Get fills dest argument with all current items if the cache.
+func (c *Cache) Get(dest interface{}) error {
+	if reflect.TypeOf(dest) != c.typ {
+		return fmt.Errorf("%w: must be %v, got %v", ErrWrongType, c.typ, reflect.TypeOf(dest))
+	}
+	c.rw.RLock()
+	defer c.rw.RUnlock()
+
+	m := reflect.ValueOf(dest)
+	for k, v := range c.items {
+		m.SetMapIndex(reflect.ValueOf(k), reflect.ValueOf(v))
+	}
+	return nil
+}
+
+// Refresh removes expired items from cache, then adds current items, then trims the cache if it's length is more than specified.
+func (c *Cache) Refresh(current interface{}) error {
+	if reflect.TypeOf(current) != c.typ {
+		return fmt.Errorf("%w: must be %v, got %v", ErrWrongType, c.typ, reflect.TypeOf(current))
+	}
+
+	c.rw.Lock()
+	defer c.rw.Unlock()
+	now := time.Now()
+
+	for k, t := range c.added {
+		if now.Sub(t) > c.retain {
+			c.removedN++
+			delete(c.items, k)
+			delete(c.added, k)
+		}
+	}
+
+	m := reflect.ValueOf(current)
+	for _, k := range m.MapKeys() {
+		key := k.Interface()
+		value := m.MapIndex(k).Interface()
+		if _, ok := c.items[key]; ok {
+			c.updatedN++
+		} else {
+			c.addedN++
+		}
+		c.items[key] = value
+		c.added[key] = now
+	}
+
+	cacheItemsN := uint(len(c.added))
+	if cacheItemsN > c.sizeLimit {
+		overSize := cacheItemsN - c.sizeLimit
+		itemList := make(sortByTimeSlice, cacheItemsN)
+		i := 0
+		for k, v := range c.added {
+			itemList[i] = sortByTime{k, v}
+			i++
+		}
+		sort.Sort(itemList)
+
+		for _, item := range itemList[0:overSize] {
+			c.removedN++
+			delete(c.items, item.key)
+			delete(c.added, item.key)
+		}
+		c.l.Debugf("Cache size exceeded the limit of %d items and the oldest values were trimmed. "+
+			"Now the oldest query in the cache is of time %s",
+			c.sizeLimit, itemList[overSize].t.UTC().Format("2006-01-02T15:04:05Z"))
+	}
+	return nil
+}
+
+// Stats returns Cache statistics.
+func (c *Cache) Stats() Stats {
```

Comment:
As I see this function is used only for debug log, also this debug log is always printed after `Refresh` function. So maybe it makes sense just to move log inside `Refresh` function and make this method private.